File: examples/spectral/static_thermal/gpu_none.cpp
// =============================================================================
// examples/spectral/static_thermal/gpu_none.cpp
//
// spectral | static thermal | single GPU (cuBLAS/cuSPARSE) | full Hilbert (no symmetry)
//
// static thermodynamic averages at a single T via qed.thermal. Twin: examples/spectral/static_thermal/gpu_none.py
//
// Requires: WITH_CUDA build + a visible CUDA device
// =============================================================================

#include "../../_shared/common.h"

#include <ed/api.h>

#include <iomanip>
#include <iostream>

int main(int argc, char** argv) {
    auto guard = ed_example::mpi_guard(argc, argv);
    constexpr std::uint64_t N = 8;
    (void)guard;

    auto op   = ed_example::heisenberg_chain(N, /*pbc=*/true);
    auto spec = ed_example::in_memory_spec(std::move(op), N);

    ed::api::ThermalOptions opts;
    opts.method      = "FTLM";
    opts.device      = "gpu";
    opts.T_min       = 0.5;
    opts.T_max       = 2.0;
    opts.num_T       = 4;
    opts.num_samples = 8;
    opts.random_seed = 0;

    auto result = ed::api::thermal(std::move(spec), opts);

    const auto& T  = result.thermo.temperatures;
    const auto& E  = result.thermo.energy;
    const auto& Cv = result.thermo.specific_heat;

    std::cout << std::fixed << std::setprecision(4);
    ed_example::rank0_print("gs_E    = ", result.ground_state_energy, "\n");
    if (!T.empty()) {
        const std::size_t iT = T.size() / 2;
        ed_example::rank0_print(
            "T_probe = ", T[iT], "  E = ", E[iT], "  Cv = ", Cv[iT], "\n");
    }

    // === Expected output (deterministic; captured on the CI reference runner) ===
    // (filled in by refresh_expected_output.py once the CPU binaries are built)
    // ===========================================================================
    return 0;
}

File: examples/spectral/ground_state_dssf/gpu_none.py
"""spectral | ground-state DSSF | single GPU (cuBLAS/cuSPARSE) | full Hilbert (no symmetry)

dynamical structure factor S_zz(omega) at three omega points. Twin: ``examples/spectral/ground_state_dssf/gpu_none.cpp``.

Requires: WITH_CUDA build + a visible CUDA device
"""
from __future__ import annotations

import os
import sys

import numpy as np

sys.path.insert(0, os.path.join(os.path.dirname(__file__), "..", ".."))
from _shared.common import heisenberg_chain, rank0_print

import qed

N = 8
H = heisenberg_chain(N, pbc=True)

obs = qed.Operator(N, 0.5)
for i in range(N):
    obs.add_one_body(qed.OP_SZ, i, 1.0)

omega = np.linspace(-5.0, 5.0, 11)
result = qed.spectral(
    H,
    [obs],
    method="ground_state_cf",
    omega=omega,
    eta=0.1,
    krylov_dim=80,
    device="gpu",
    verbose=False,
)

mid = len(result.omega) // 2
rank0_print(f"S(w=-5.0) = {result.S_real[0]:.6f}")
rank0_print(f"S(w= 0.0) = {result.S_real[mid]:.6f}")
rank0_print(f"S(w= 5.0) = {result.S_real[-1]:.6f}")

# === Expected output (deterministic; captured on the CI reference runner) ===
# (filled in by refresh_expected_output.py once the CPU binaries are built)
# ===========================================================================
